PR for Review of Invoke-based Rewrite

jseppi opened the pull request:
Followed this guide to make a PR with all the changes so that a GitHub review can be done: http://astrofrog.github.io/blog/2013/04/10/how-to-conduct-a-full-code-review-on-github/

dhcole committed:
Docker image and build script

dhcole committed:
Add README

dhcole committed:
Update README

dhcole committed:
Include scripts for cloning from GitHub, building with Jekyll, and uploading to S3

dhcole committed:
Install Gemfile dependencies if present

dhcole committed:
Use bash for shell

dhcole committed:
Support multiple generators for build (hugo is a stub)

dhcole committed:
Set not read variable

dhcole committed:
Comment about directory redirects being slow

dhcole committed:
Fix var. Post to endpoint on any exit

dhcole committed:
Merge pull request #2 from 18F/pull/full-build

Add more steps to the build process

dhcole committed:
Use Ubuntu base image. Simplify scripts.

dhcole committed:
Preload Jekyll versions

dhcole committed:
Install github-pages gem

dhcole committed:
Remove callback env var

dhcole committed:
If a source build, clone the source and push to destination

jeremiak committed:
Merge pull request #3 from 18F/pull/clone

Adds support for cloning builds

dhcole committed:
Add commit details to site object

jeremiak committed:
Merge pull request #4 from 18F/pull/commit-details

Add commit details to site object

dhcole committed:
Fix conditional check (should be not null instead of null)

dhcole committed:
Merge pull request #5 from 18F/temp

Fix conditional check (should be not null instead of null)

dhcole committed:
Update CONTRIBUTING.md

Commit:
Remove binary characters that crash node

@yozlet was having some issues adding one of his
repositories to federalist and it turns out that
some of the output from the build process was colored,
and those color characters were causing the JSON
parser to crash in node. This resulted in bad
communication between the builder and the main
application.

Now output should be scrubbed using `tr`

Commit:
Update documentation to include steps on local dev

Commit:
Base64 encode message

Instead of using character replacement this base64 encodes the message
and removes newlines in the output

Commit:
Use 4xx error pages if they are set up for Jekyll

Using the aws s3 website command
(http://docs.aws.amazon.com/cli/latest/reference/s3/website.html)
to set 404 error page if it is present.

jeremiak committed:
Merge pull request #7 from 18F/use-4xx-error-pages

Use 4xx error pages if they are set up for Jekyll

jeremiak committed:
Revert "Use 4xx error pages if they are set up for Jekyll"

jeremiak committed:
Merge pull request #9 from 18F/revert-7-use-4xx-error-pages

Revert "Use 4xx error pages if they are set up for Jekyll"

el-mapache committed:
Merge pull request #6 from 18F/troubleshooting

Whitelist ASCII characters for status update POST

Commit:
Change the `mv` command to exclude the directory

Currently, the build.sh command makes a `_site` directory, and
then proceeds to try to move everything in the current directory
into `_site`, *including the `_site` itself*.

This commit instead moves everything *except* `_site` into site.
The use of the `shopt` command is to find hidden files (because
strangely not including this causes bash to freak out a bit).

This could certainly be more elegant, but seems to work.

Commit:
Move shopt outside of if block

In my testing, it looks like shopt will not work if it is
within an if/else/then block. Huge h/t to @toolness for
his help on figuring this one out.

jeremiak committed:
Merge pull request #12 from 18F/fix-static-site-build

Fix static site build

Commit:
Use nvm and npm if respective files are present

First attempt at getting nvm and npm to run on Federalist projects.

I don’t think this is the _best_ solution because the output of these
commands won’t end up in the callback payload and therefore will make
it hard to debug

el-mapache committed:
Dont pipe output to /dev/null

jmhooper committed:
Add callback to federalist builder

This commit adds a callback to Federalist Builder so the container can let Federalist Builder know that it has finished building and is available.

This commit also tells the container to sleep after it has finished running so that it will not try to restart and run the build again.

jmhooper committed:
Use python3 to build awscli

This commit uses `python3` to install awscli since `python` is no longer installed in `18fgsa/docker-ruby-ubuntu`.

brandocalrissian committed:
Merge pull request #17 from 18F/jmhooper-python3

Use python3 to build awscli

jmhooper committed:
Only set AWS_DEFAULT_REGION if it is not already set

The `AWS_DEFAULT_REGION` is the variable that describes which region the site will be published to.

Before this commit, the region was hardcoded to "us-east-1".
This created problems migrating to GovCloud.

Requiring the AWS region to be explicitly set invites its own problems because it requires us to modify the production app running in E/W.

This commit configures the build environment so the AWS region can be overriden, but defaults to "us-east-1".
This allows us to support legacy behavior, while being able to provide an override to publish sites to S3 in GovCloud.

closes #19

brandocalrissian committed:
Merge pull request #20 from 18F/jmhooper-change-aws-region

Only set AWS_DEFAULT_REGION if it is not already set

jmhooper committed:
Use login mode with bash invocation

This commit use bash with login mode so that `nvm` will be loaded from `~/.profile`.

brandocalrissian committed:
Merge pull request #21 from 18F/jmhooper-use-login-mode

Use login mode with bash invocation

jmhooper committed:
Send request to builder callback endpoint even if main callback fails

We are having issues on staging with the main callback failing which is preventing the builder callback from being called.
This commit fixes that by not exiting and going straight to the builder callback even if the main callback fails.

jmhooper committed:
Rename PREFIX to SITE_PREFIX

This commit responds to a rename of the variable `PREFIX` to
`SITE_PREFIX` in the container environment. This rename was necessary to
avoid a collision with the `PREFIX` environment variable used by npm.

ref 18F/federalist#584

jmhooper committed:
Rename PREFIX to SITE_PREFIX

This commit responds to a rename of the variable `PREFIX` to
`SITE_PREFIX` in the container environment. This rename was necessary to
avoid a collision with the `PREFIX` environment variable used by npm.

ref 18F/federalist#584

jmhooper committed:
Don't load nvm in build script

This commit does not load the nvm function in build.sh. This is not
necessary at this point since we invoke these scripts in login mode, so
nvm should be available.

brandocalrissian committed:
Merge pull request #23 from 18F/jmhooper-change-prefix

Rename PREFIX to SITE_PREFIX

brandocalrissian committed:
Merge pull request #13 from 18F/jk-use-nvm

Use nvm and npm if respective files are present

jmhooper committed:
Initialize nvm before build script

This commit reverts changes in 280b15961363cbae4dbe3b39f2016e08ccdd96eb
and initializes nvm before using it in the build script.

Even though using bash in login mode makes nvm available in login mode to the
script it is invoked on, it does not make it available to child scripts
that script invokes. Because of this, login mode cannot be used to make
nvm available to build.sh. This change removes login mode and
initializes nvm in build.sh before using it.

brandocalrissian committed:
Merge pull request #24 from 18F/jmhooper-init-nvm

Initialize nvm before using it build script

jmhooper committed:
Add build container logging

This commit communicates with the web application every time it finishes
running a script. The purpose of this is to provide the web application
with build logs that it can display to the user to be used for more
in-depth debugging when issues happen in the build container.

brandocalrissian committed:
Merge pull request #25 from 18F/jmhooper-build-container-logging

Add build container logging

jmhooper committed:
Remove explicit source option from Jekyll build

This commit removes the explicit source option from the Jekyll config.

The CLI option overrides any option specified in `_config.yml`. This meant sites that have their site's source files in a directory besides the project root could not be built correctly.

Ref 18f/federalist#180

jmhooper committed:
Merge pull request #29 from 18F/jmhooper-remove-source-cli-option

Remove explicit source option from Jekyll build

jmhooper committed:
Enable encryption on objects added to the bucket

This commit configures the AWS cli to use server side encryption on
objects added to S3. This should not result in any noticeable changes
for sites, but will mean that the data is encrypted in S3's backend.

brandocalrissian committed:
Merge pull request #30 from 18F/jmhooper-encryption

Enable encryption on objects added to the bucket

jmhooper committed:
Only install production node modules

To keep the amount of time `npm install` takes, the amount of disk it
uses, and the amount of native packages it depends on, this commit adds
the `--only=production` flag to `npm install` so that it will only
install packages listed as dependencies.

jmhooper committed:
Add python 2.7 to the build container

Node-gyp requires Python 2.7 to build. Lots of older node modules depend
on node-gyp. This commit makes Python 2.7 available in the build
container for those older node modules.

brandocalrissian committed:
Merge pull request #31 from 18F/jmhooper-install-production-node-modules

Only install production node modules

brandocalrissian committed:
Merge pull request #32 from 18F/jmhooper-python-2.7

Add python 2.7 to the build container

jmhooper committed:
Fail build if any of the build steps

This commit causes a build to fail if one of the build steps fails. This happened previously, but was removed in a4c5a3d52e5f3a4675e52e63ec72f22b395f0d97 to debug.

This commit reintroduces the feature. It also includes some logging that will be helpful for determining why failures happen from the cloud.gov logs.

Ref #784

jmhooper committed:
Use `--production` instead of `--only=production` on npm install

The `npm install` option `--only=production` is not supported by npm v4
which is our default node version. This commit switches to
`--production` which looks like it is supported by v4, v6, and v7.

brandocalrissian committed:
Merge pull request #35 from 18F/jmhooper-production-install

Use `--production` instead of `--only=production` on npm install

brandocalrissian committed:
Merge pull request #34 from 18F/jmhooper-error-fail

Fail build if any of the build steps error

jmhooper committed:
Remove git details from site config

Prior to this commit, details about the commit were added to the site's
config before the site was built. This created problems when the
commit's message had syntax that resulted in malformed yaml, for
example, a `"` character in a commit message. This would create yaml
parse errors when the site was building, failing the build.

This commit removes that commit data since it does not seem to be used
by any sites, and there doesn't seem to be a straightforward fix for the
issue above:

Ref #823

brandocalrissian committed:
Merge pull request #36 from 18F/jmhooper-remove-git-data

Remove git details from site config

jmhooper committed:
Stop scary "missing federalist script" warning

When a build a `pacakge.json` file, the build container would run `npm run federalist || true`. This allowed users to add custom npm scripts, but had a few problems:

- Sites without a `federalist` script would see a disconcerting npm error about a missing federalist script
- Sites with a broken `federalist` script would not see their builds fail

This commit adds a conditional that checks whether the script is present in the `package.json` before attempting to run it. This means that the error will no longer appear for sites w/o a federalist script. In addition, the build will fail if the federalist script has a nonzero exit code.

Ref #33

brandocalrissian committed:
Merge pull request #37 from 18F/jmhooper-conditional-federalist-script

Stop scary "missing federalist script" warning

jmhooper committed:
Improve build container logging and remove superfluous activities

This commit is the result of work to debug an issue that was noticed on staging where federalist-builder was reporting that all builds were timing out for some larger sites. The issue was very hard to discover due to the lack of visibility into the container's activities.

This commit adds additional visibility, and removes several superfluous activities that were discovered to be the cause of the mysterious build timeouts. The changes are:

- Use `tee` to follow logs from build scripts in real time in container logs
- Add logging statements to make it clear where in the process the build is
- Use a base64 encoded string for the output in the build logs callback to avoid weird javascript string parsing errors
- Limit the length of build log output when calling back to the log callback to keep excessively long logs from exceeding bash argument length
- Preserve newlines in callbacks back to Federalist
- Remove redirect creation in `publish.sh` since it is long-running and unnecessary in an S3 bucket with a propper website configuration

brandocalrissian committed:
Merge pull request #38 from 18F/jmhooper-fix-container-errors

Improve build container logging and remove superfluous activities

jmhooper committed:
Allow sites to use RVM to swtich ruby versions

This commit adds the ability for sites to set their ruby version using a `.ruby-version` file.

This commit adds a Federalist user to build the sites since running RVM as root can be kind of shaky. It also moves more of the configuration for the ruby install into build.sh. This is necessary because switching ruby versions means the old configurations will be left behind with the old version.

Ref #6

jmhooper committed:
Unset pipefail before using pipes that can fail

We have several points where we do something like this:

```shell
curl some.url.gov ; curl some.other.url.gov || true
```

The purpose of this is to not break the build if any external services are down. This is very useful when reporting build logs, since the request is sent in the middle of the build process.

Additonally, we have build steps where we use `tee` to output the build output into the logs. That looks like this:

```shell
/app/build.sh 2>&1 | tee /dev/stderr
```

When `/app/build.sh` fails there, we want the entire build to fail.

The result of the above is that we want want the container to have `pipefail` enable some of the time, and disabled other times. This commit sets it when the build starts, and momentarily disables it when we do not want it.

brandocalrissian committed:
Merge pull request #39 from 18F/jmhooper-rvm

Allow sites to use RVM to swtich ruby versions

brandocalrissian committed:
Merge pull request #40 from 18F/jmhooper-unset-pipefail

Unset pipefail before using pipes that can fail

jmhooper committed:
Fix issue where timed out builds reported success.

This commit fixes an issue that arose with builds that timed out. When builds would time out, cloud.gov would stop the build would would trigger a call to the `post`. Since the exit code was zero, this would report to Federalist that the build completed successfully, even though it timed out.

This commit changes the `main.sh` script to detect when a build does not complete before exiting, and reports an error in that case, with an error message hinting that the build may have timed out.

jmhooper committed:
Build site as root user

This commit modifies the build container so that it builds the site as
the root user. This was necessary because, while it appears that using a
`federalist` user works locally, it does not appear to work in
cloud.gov.

brandocalrissian committed:
Merge pull request #41 from 18F/jmhooper-fix-timeout-bug

Fix issue where timed out builds reported success.

brandocalrissian committed:
Merge pull request #42 from 18F/jmhooper-use-root-user

Build site as root user

jmhooper committed:
Don't publish dotfiles

After build, site's may contain files like .travis.yml, .eslintrc, or a `.git` directory which we don't want to appear on the public internet. This commit removes those before publishing.

jmhooper committed:
Fix dotfile regex

The regex we were using to match dotfiles didn't work because it matched
`.` and `..`. This commit users a more sophisticated regex to look for
dotfiles.

jmhooper committed:
Timeout builds after they run for 20 minutes

This commit adds a new script named `run.sh` which invokes `main.sh`
with an instruction to time it out after 20 minutes. This means that if
the process building the site takes longer than 20 minutes to execute,
this new script will stop it.

After setting the timeout, the run script then sleeps, and invocation
that was removed from the main script.

The introduction of a new script was necessary because `timeout` cannot
be used to run a function. Complex commands must be contained inside a
separate script and run that way.

brandocalrissian committed:
Merge pull request #46 from 18F/jmhooper-timeout

Timeout builds after they run for 20 minutes

jmhooper committed:
Only publish files that have changed since the last publish

This commit replaces `publish.sh` with a slightly more sophisticated ruby script in `publish.rb`. This script downloads the MD5 hashes of the objects in the buckets and compares them to the MD5 hashes of the local files before uploading. It uses this information to determine if a file has changed since the files were last published, and only re-uploads files that have a different checksum.

This commit also create redirect objects for pages with a path in the form of `/path/to/page/`, so that `/path/to/page` will redirect to the correct place concerning the base url.

Ref #927

brandocalrissian committed:
Merge pull request #47 from 18F/jmhooper-smarter-publish

Only publish files that have changed since the last publish

jmhooper committed:
Add baseurl to root redirect destination

The root URL's redirect was not prefixed by the root url, breaking the sites built without a custom domain. This commit adds the baseurl prefix to the redirect for the root url.

Ref #53

jmhooper committed:
add hugo to docker image and to build script

jmhooper committed:
Merge pull request #54 from 18F/jmhooper-add-baseurl-to-destination

Add baseurl to root redirect destination

jseppi committed:
Merge pull request #51 from 18F/add-hugo-support

Add hugo to docker image and to build script

jseppi committed:
Add link to deployment instructions

jseppi committed:
Merge pull request #55 from 18F/50-link-deploy-instructions

Add link to deployment instructions

Commit:
Update hugo download URL and version to 0.23

jseppi committed:
Merge pull request #57 from 18F/update-hugo-bin-url

Update hugo download URL and version to 0.23

Commit:
update to reflect new name of repo and image

jseppi committed:
Merge pull request #61 from 18F/readme-new-repo-name

Update README.md to reflect new name of repo and image

jseppi committed:
Change build engine language a bit to include hugo

jseppi committed:
Merge pull request #64 from 18F/add-hugo-readme

Change build engine language a bit to include hugo

jseppi committed:
Fix README header

jmhooper committed:
Merge pull request #66 from 18F/65-readme-header

Fix README header

Commit:
use correct filename for publish log output

jseppi committed:
Merge pull request #68 from 18F/59-log-publish.rb

use correct filename for publish log output

Commit:
Fix logging of template repo push to owner github

jseppi committed:
Merge pull request #69 from 18F/62-push-repo-name

Fix logging of template repo push to owner github

jseppi committed:
Check that "scripts" exists in package.json

Ref https://github.com/18F/federalist/issues/1117

jseppi committed:
Merge pull request #71 from 18F/check-scripts-exists

Check that "scripts" exists in package.json

jseppi committed:
Only include `--baseURL` arg when `BASEURL` exists

and is non-empty.

Ref #72

jseppi committed:
Merge pull request #73 from 18F/72-fix-hugo-baseurl

Only include `--baseURL` arg when `BASEURL` exists

jseppi committed:
minor fix

jseppi committed:
Merge pull request #75 from 18F/jseppi-patch-1

Add the rest of the word "repository"

Commit:
[WIP] Pythonizing using pyinvoke

Commit:
create the publishing logic

Commit:
refactor source files

Commit:
pylint cleanups

Commit:
rm TODO comment

Commit:
refactor, cleanup, fix PYTHONIOENCODING bug

Commit:
fix hugo dl and path

Commit:
fix ruby-version use; more time logging

Commit:
add LANG env var to build_env

Commit:
better env sample template

Commit:
properly call main task

Commit:
dev reqs, timeout, logging reorg, etc.

Commit:
add basic circleci config

Commit:
badges

Commit:
upgrade colorlog

Commit:
add bandit step

Commit:
simplify circle docker testing

Commit:
add run.sh to sleep after build

Commit:
comment cleanup

Commit:
revert logging; fix json log payload

Commit:
more dev info

Commit:
rm unused file

Commit:
use env for sensitive args

Commit:
logging enhancements and cleanup

Commit:
rm colorlog

Commit:
rm erroneous version output

Commit:
add more info to initial main log

Commit:
fix some encoding and logging issues

Commit:
more log cleanup, use absolute /tmp dir for builds

Commit:
flake8 cleanup

Commit:
add codeclimate

Commit:
minor SITE_PREFIX addendum

brendansudol commented on log_utils/remote_logs.py:
```diff
@@ -0,0 +1,112 @@
+    '''
+    # Replicates log_output() function in main.sh
+    if len(output) > limit:
+        output = limit_msg
```
at some point (in a future PR), if the output length is too long, perhaps we should send a truncated output (i.e., `output[:limit] + '...'`)? it seems like that may be more helpful than simply the `limit_msg`

jseppi replied:
yup, great idea. This was just a port of the existing logic. Opened https://github.com/18F/federalist-garden-build-py/issues/6 to track.

brendansudol commented on publishing/SiteObject.py:
```diff
@@ -0,0 +1,197 @@
+import mimetypes
+import hashlib
+import gzip
+import binascii
```
i tend to alphabetize imports within these groups for readability; this isn't a pep8 rule, simply a personal pref (no change necessary here)

brendansudol commented on publishing/SiteObject.py:
```diff
@@ -0,0 +1,197 @@
+        filename = self.filename
+        if self.dir_prefix:
+            filename = remove_prefix(filename,
+                                     path.join(self.dir_prefix, ''))
```
what's the purpose of `path.join(self.dir_prefix, '')` vs. just `self.dir_prefix`? just to add a `/` if the dir_prefix doesn't have one already?

jseppi replied:
yup, exactly

brendansudol commented on publishing/s3publisher.py:
```diff
@@ -0,0 +1,180 @@
+                                         s3_client=s3_client)
+
+    # Make dicts by filename of local and remote objects for easier searching
+    remote_objects_by_filename = {}
```
this could be consolidated a bit (if you'd like):

```
remote_objects_by_filename = {
    (path.join(directory, obj.filename) if obj.filename else directory): obj
    for obj in remote_objects
}
```

jseppi replied:
nice, thanks

brendansudol commented on publishing/s3publisher.py:
```diff
@@ -0,0 +1,180 @@
+            modified_objects.append(local_obj)
+
+    # Create a list of the remote objects that should be deleted
+    deletion_objects = []
```
another place where list comprehension saves some letters:

```
deletion_objects = [
    obj for filename, obj in remote_objects_by_filename.items()
    if not local_objects_by_filename.get(filename) 
]
```

brendansudol commented on tasks/main.py:
```diff
@@ -0,0 +1,281 @@
+    for priv_val in private_values:
+        text = text.replace(priv_val, replace_string)
+
+    return text
```
if you want to be cool and functional... ;)

`reduce(lambda txt, val: txt.replace(val, replace_string), private_values, text)`

(though this is harder to read than yours imo)

jseppi replied:
I would absolutely do it that way in javascript, but I shy away from the functional stuff in python because of the syntax and silly `lambda` keyword

brendansudol commented on publishing/s3publisher.py:
```diff
@@ -0,0 +1,180 @@
+            file.upload_to_s3(bucket, s3_client)
+
+            delta = datetime.now() - start_time
+            LOGGER.info(f'Uploaded {file.s3_key} in '
```
can use your `delta_to_mins_secs` method here

jseppi replied:
I thought about that, but these uploads are often on the order of fractions of a second

brendansudol reviewed (approved):
this is a great and comprehensive initial rewrite to python; there's room for refinement down the road but i vote to ship it!

jseppi commented:
Awesome, thanks for the review, @brendansudol!

jseppi commented:
@jmhooper you want to give it a once-over as well?

Commit:
improve output truncation, minor code clean-ups

Instead of totally throwing away truncated output, instead
keep most of it up to a limit (50,000 chars) and add a truncation message.
Closes #6.

Also some minor code cleanups per @brendansudol's review.

Commit:
add cov args to pytest call

and simplify docker tag name in circleci config

jseppi committed:
Merge pull request #7 from 18F/add-coverage

add cov args to pytest call

Commit:
docker-compose in circle; send coverage to CC

jseppi committed:
Merge pull request #8 from 18F/circleci-docker-compose

docker-compose in circle; send coverage to CC

Commit:
update README

Commit:
config coverage reporting with .coveragerc

omit test/ dir from coverage reports since that's where tests themselves are

Commit:
more readme updates

jseppi committed:
Merge pull request #9 from 18F/update-readme

Update README

jmhooper reviewed (approved):
This comes through a lot better than the bash 💯. I like it.

jseppi commented:
![image](https://user-images.githubusercontent.com/697848/32754240-a96741ac-c895-11e7-8ecf-acf815ec3cc2.png)

jseppi commented:
Thanks friends, going to go ahead and close this review PR. 🎉
